### ISSUE TITLE
anndata reading error with version 0.8.0

### ISSUE THREAD

**Issue by cornhundred** [scverse/anndata#731]
Hi, 

Scanpy's `sc.read_h5ad` method appears to be having an error in how it is reading our AnnData HDF5 file when we use the new 0.8.0 version of the AnnData library:

Loading using anndata version 0.7.8
```
AnnData object with n_obs × n_vars = 121393 × 649
    obs: 'Custom cells sets', 'Datasets', 'volume', 'center_x', 'center_y'
```

Loading using anndata version 0.8.0
```
AnnData object with n_obs × n_vars = 121393 × 649
    obs: 'Custom cells sets', 'Datasets', '__categories', '_index', 'center_x', 'center_y', 'volume'
    var: '_index'
```

I can provide a Google Colab example soon to reproduce.

**Comment by ivirshup** [comment]
> I can provide a Google Colab example soon to reproduce.

That would be very useful, thanks!

**Comment by cornhundred** [comment]
Hi @ivirshup, I went ahead and used Kaggle to make the example notebooks since our AnnData exceeds Github file size requirements. The kaggle notebooks reproduce our issues. Let me know if you have any issues running the notebook. 

* https://www.kaggle.com/cornhundred/test-anndata-read-version-0-7-8

* https://www.kaggle.com/cornhundred/test-anndata-read-version-0-8-0

**Comment by ivirshup** [comment]
Thanks! I can confirm the bug.

Any chance you know what version of anndata wrote this file?

**Comment by ivirshup** [comment]
As a workaround for right now, if you write what you've read in with `anndata==0.7.8` the new release will be able to read that.

**Comment by cornhundred** [comment]
Thanks @ivirshup, so I can re-write the anndata and then read it correctly? Would you be able to fork the kaggle notebook to demonstrate?

**Comment by cornhundred** [comment]
> Thanks! I can confirm the bug.
> 
> Any chance you know what version of anndata wrote this file?

Also, this particular file was generated using the HDF5 C++ library https://support.hdfgroup.org/ftp/HDF5/releases/COPYING (not sure of version).

**Comment by ivirshup** [comment]
> Thanks ivirshup, so I can re-write the anndata and then read it correctly?

Not sure how I would do it through kaggle, but I downloaded the dataset and can show you with some conda environments.

```sh
$ mamba create -yn anndata-0.7.8 -c conda-forge python=3.9 "anndata==0.7.8"
...
$ mamba create -yn anndata-0.8.0 -c conda-forge python=3.9 "anndata==0.8.0"
...
$ conda run -n anndata-0.7.8 python3 -c "import anndata; anndata.read_h5ad('brain_test.hdf5').write_h5ad('brain_test_rewritten.h5ad')"
$ conda run -n anndata-0.8.0 python3 -c "import anndata; print(anndata.read_h5ad('brain_test_rewritten.h5ad'))"
AnnData object with n_obs × n_vars = 121393 × 649
    obs: 'Custom cells sets', 'Datasets', 'volume', 'center_x', 'center_y'
```

> Also, this particular file was generated using the HDF5 C++ library

So, was this not written by this library? I.e. this file was not written by the python anndata implementation?

That would make sense, since this would otherwise have needed to be written by an in-development version of `anndata`. Because I'm pretty sure there was never a release of a version of anndata that would have written this. In `anndata==0.7.0rc1` dataframes should have had the encoding metadata tags ([implementation](https://github.com/theislab/anndata/blob/142e6d02deeb14e5c71a7ae7d7dcced8cf73da97/anndata/_io/h5ad.py#L235-L254)), while in `anndata==0.6.22.post1` dataframes were written as a single rec_array ([implementation](https://github.com/theislab/anndata/blob/75296bb210d141fa1de24960412bfcab77a9604d/anndata/readwrite/write.py#L224-L245)).

**Comment by cornhundred** [comment]
Thanks @ivirshup, yes I didn't realize this Anndata file was not written by Anndata but by a C++ HDF5 library in a custom manner. We hadn't run into the issue loading it into Anndata until this relese, so for now we'll stick to using the 0.7.8 to load these versions of the AnnData. I'll go ahead and close the issue since this is an issue we're having with files that weren't written by AnnData.

**Comment by ivirshup** [comment]
I think I might be okay with continuing to support this. I had explicitly removed support for files like this in 0.8.0 because I thought I would be the only person with any files in this format on my machine 😆 

It’s a very easy fix on our side, though it'd probably end up throw a user visible warning saying to rewrite the file.

I would like to hear more about your C++ library for working with anndata files. I would especially like to make sure implementations don’t diverge and keep compatibility. Any chance this code publicly available or we could talk about it offline?

**Comment by cornhundred** [comment]
Thanks that sounds good @ivirshup I'll email you.

**Comment by NataliaDSmirnova** [comment]
> I think I might be okay with continuing to support this. I had explicitly removed support for files like this in 0.8.0 because I thought I would be the only person with any files in this format on my machine 😆
> 
> It’s a very easy fix on our side, though it'd probably end up throw a user visible warning saying to rewrite the file.
> 

It would be very helpful if you implement the fix because there are some users that have already exported Anndata before we applied the fix to our application.

**Comment by flying-sheep** [comment]
@ivirshup is this still relevant? We’re at 0.9 now.

If this still needs a fix, we should think of #747

**Comment by flying-sheep** [comment]
I’m going to assume it isn’t. Please comment to reopen in case it still applies.
